### PR TITLE
Perfscore dumping/dasm improvements

### PR DESCRIPTION
**Pull request by briansull**
Added support for all currently implemented AVX2, SSE2, etc.. instructions

### CODE REVIEW

**Comment by briansull** [comment]
@dotnet/jit-contrib PTAL

**Comment by CarolEidt** [comment]
Still looking, but it might have been nice to separate out the changes where you add the block weight and perf score to the instruction group.

File: src/coreclr/src/jit/emitxarch.cpp [hunk 1]
```diff
@@ -13703,6 +13703,10 @@ emitter::insFormat emitter::getMemoryOperation(instrDesc* id)
 
 #if defined(DEBUG) || defined(LATE_DISASM)
 
+#ifndef add
+#define add(a, b) ((a) + (b))
```
Comment (CarolEidt) [inline src/coreclr/src/jit/emitxarch.cpp]: Why is this needed? Seems a pretty weird place to put this.

**Comment by CarolEidt** [comment]
@briansull - what reference are you using for the instruction throughput and latency info? It would be good to add that to the function header comment for `emitter::getInsExecutionCharacteristics`

**Review by CarolEidt — COMMENTED** [review]
The structural changes seem reasonable, but it would be good to be able to review the numbers relative to the source you're using.

**Comment by briansull** [comment]
I am using the Intel Skylake-X data from  Agner and https://uops.info/
https://www.agner.org/optimize/instruction_tables.pdf

I will add this information to the function header

File: src/coreclr/src/jit/emitxarch.cpp (hunk 1 above)
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: Some places I use max(a,b) to compute the instruction Latencies and now I see that that some instructions should instead compute the latency using  add(a,b).

I could rewrite these using a + b, if you feel strongly.

Comment (echesakovMSFT) [reply src/coreclr/src/jit/emitxarch.cpp]: In my opinion, we shouldn't create macros for simple arithmetic operations.

Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: OK I will change these

File: src/coreclr/src/jit/emitxarch.cpp [hunk 2]
```diff
@@ -14282,6 +14295,26 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
...
+        case INS_movd:
+            if (memAccessKind == PERFSCORE_MEMORY_NONE)
+            {
+                // movd   r32, xmm   or  xmm, r32
+                result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+                result.insLatency    = PERFSCORE_LATENCY_3C;
+            }
+            else if (memAccessKind == PERFSCORE_MEMORY_READ)
+            {
+                // movq   xmm, m32
```
Comment (CarolEidt) [inline src/coreclr/src/jit/emitxarch.cpp]: Should this (and below) be `movd`?

File: src/coreclr/src/jit/emitxarch.cpp [hunk 3]
```diff
@@ -14282,6 +14295,26 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
             result.insLatency    = PERFSCORE_LATENCY_1C;
             break;
 
+        case INS_movd:
+            if (memAccessKind == PERFSCORE_MEMORY_NONE)
+            {
+                // movd   r32, xmm   or  xmm, r32
+                result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+                result.insLatency    = PERFSCORE_LATENCY_3C;
```
Comment (CarolEidt) [inline src/coreclr/src/jit/emitxarch.cpp]: As I read the document, this should be `PERFSCORE_LATENCY_2C`

**Comment by briansull** [comment]
Here is a sample using the new disassembly output information
Diff

**Comment by briansull** [comment]
Updated
@dotnet/jit-contrib PTAL

File: src/coreclr/src/jit/emitxarch.cpp [hunk 4]
```diff
@@ -14447,156 +14530,319 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
...
+        case INS_pmovsxwq:
+        case INS_pmovsxdq:
+        case INS_pmovzxbw:
+        case INS_pmovzxbd:
+        case INS_pmovzxbq:
+        case INS_pmovzxwd:
+        case INS_pmovzxwq:
+        case INS_pmovzxdq:
+            result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+            result.insLatency += PERFSCORE_LATENCY_3C;
```
Comment (CarolEidt) [inline src/coreclr/src/jit/emitxarch.cpp]: It seems that these have unit latency for xmm registers?

**Comment by BruceForstall** [comment]
So every code change to a block will lead to 2 diffs: one (or more) for the code changes and one for the perf score diff. Hopefully that's not too distracting, and is useful in most cases.

**Review by CarolEidt — COMMENTED** [review]
I've just done a sample review of the latency & throughput numbers relative to SkylakeX according to the referenced instruction_tables.pdf, but there were some things that seemed to be discrepancies.

File: src/coreclr/src/jit/emitxarch.cpp [hunk 5]
```diff
@@ -13898,8 +13891,28 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
...
+                regNumber indexReg = id->idAddr()->iiaAddrMode.amIndxReg;
+                if (indexReg != REG_NA)
+                {
+                    regNumber baseReg = id->idAddr()->iiaAddrMode.amBaseReg;
+                    if (baseReg != REG_NA)
+                    {
+                        ssize_t dsp = emitGetInsAmdAny(id);
+                        if (dsp != 0)
+                        {
+                            result.insThroughput = PERFSCORE_THROUGHPUT_1C; // three components
```
[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: I don't believe that a three component  LEA has a latency of 3, in Agner
I have used the instruction timings from from https://uops.info/table.html 
 LEA_BID (R64)  https://uops.info/html-instr/LEA_BID_R64.html

[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: In any case the issue rate of LEA  (0.5 or 1.0) is always worse than the add instruction (0.25), so we should avoid using LEA instructions.

[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: I found this info on LEA on the llvm.org  site:

https://reviews.llvm.org/D32277

According to Intel's Optimization Reference Manual for SNB+:
" For LEA instructions with three source operands and some specific situations, instruction latency has increased to 3 cycles, and must
  dispatch via port 1:
- LEA that has all three source operands: base, index, and offset
- LEA that uses base and index registers where the base is EBP, RBP,or R13
- LEA that uses RIP relative addressing mode
- LEA that uses 16-bit addressing mode "

I can change the latency to 3 cycles, for the three operand case, and can open an issue to migrate away from using LEA instruction.

File: src/coreclr/src/jit/emitxarch.cpp (hunk 3 above)
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: OK, I can change this,

File: src/coreclr/src/jit/emitxarch.cpp (hunk 2 above)
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: Yes, cut and paste  issue

File: src/coreclr/src/jit/emitxarch.cpp [hunk 6]
```diff
@@ -14282,6 +14295,26 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
...
+            if (memAccessKind == PERFSCORE_MEMORY_NONE)
+            {
+                // movd   r32, xmm   or  xmm, r32
+                result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+                result.insLatency    = PERFSCORE_LATENCY_3C;
+            }
+            else if (memAccessKind == PERFSCORE_MEMORY_READ)
+            {
+                // movq   xmm, m32
+                result.insThroughput = PERFSCORE_THROUGHPUT_2X;
```
[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: This initial latencies are set above (See line 13733 -13795) where we model the memory latency

[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: Typically you will see a blank value in Agner when the instruction has a memory latency.
That means the memory latency is dominate.

In my model of the memory latency, if I can (easily) prove that the read/write is to the stack I use a 
PERFSCORE_LATENCY_RD_STACK value, which is one cycle less than PERFSCORE_LATENCY_RD_WR_GENERAL.
as it is more likely to be in the L0 cache.

[the remark this answers is not in the document]
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: I can add an max(a,b) here as I do in other place where the memory latency may be greater than a value specified by Agner.

File: src/coreclr/src/jit/emitxarch.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/coreclr/src/jit/emitxarch.cpp]: That information is much more illuminating. I couldn't find a latency on either of the links you indicate, but in the instruction_tables.pdf it has a latency of 3 for "3 components in address rip relative address" (page 255). I find that table a bit confusing because it's not clear whether the lack of a latency in a row implies that it's the same as above, or simply that it's not provided (like in the case of the rip relative address).

File: src/coreclr/src/jit/emitxarch.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/coreclr/src/jit/emitxarch.cpp]: I see - thanks. Or you could just add a comment; either way.

**Review by CarolEidt — APPROVED** [review]
LGTM - thanks for the clarifications

File: src/coreclr/src/jit/emitxarch.cpp (hunk 4 above)
Comment (briansull) [reply src/coreclr/src/jit/emitxarch.cpp]: These instructions have a latency of 3 cycles according to uops.info
The Agner table omits these instructions for the Skylake-X
